locator/tablets.cc

```diff
@@ -240,15 +240,18 @@ tablet_replica_set get_new_replicas(const tablet_info& tinfo, const tablet_migra
     return replace_replica(tinfo.replicas, mig.src, mig.dst);
 }
 
-tablet_replica_set get_primary_replicas(const tablet_info& info, const tablet_transition_info* transition) {
+tablet_replica_set get_primary_replicas(const locator::tablet_map& tablet_map, tablet_id tid, std::function<bool(const tablet_replica&)> filter) {
+    const auto& info = tablet_map.get_tablet_info(tid);
+    const auto* transition = tablet_map.get_tablet_transition_info(tid);
+
     auto write_selector = [&] {
         if (!transition) {
             return write_replica_set_selector::previous;
         }
         return transition->writes;
     };
-    auto primary = [] (tablet_replica_set set) -> tablet_replica {
-        return set.front();
+    auto primary = [tid, filter = std::move(filter)] (tablet_replica_set set) -> tablet_replica {
+        return maybe_get_primary_replica(tid, set, filter).value();
```

Comment:
In theory this won't throw bad_optional_access because you disabled primary_replica on node scope, which is perfectly fine, so there should always be a replica returned within that optional.
I'm pretty sure it is possible though that if the cluster is configured with a DC that has RF=0 and the API gets called with scope=dc, your `value()` call here will throw. I think adding a check here and returning empty set is the right solution.

Reply:
The primary-replica-only should be used only with scope:all by load-and-stream.  In this case, the filter is a no-op. However, if the get_primary_replicas is used by other callers, the filter could be non-empty. Thus maybe_get_primary_replica could return nullopt.